Support `@layer` (#955)
vanilla-extract-css/vanilla-extract: site/docs/global-api/global-layer.md

```

## Assigning styles

Styles can be assigned to a layer by name, using the `@layer` key in the style object.

In this example, we first import the `layers.css.ts` stylesheet, setting up the order of the layers, then create a style within the `reset` layer.

```ts compiled
// reset.css.ts

import { style } from '@vanilla-extract/css';
import './layers.css.ts';

export const noMargin = style({
  '@layer': {
    reset: {
      margin: 0
    }
  }
});

// layers.css.ts

import { globalLayer } from '@vanilla-extract/css';

globalLayer('reset');
globalLayer('framework');
globalLayer({ parent: 'framework' }, 'typography');
```

Best practice would be to expose the layer references from the `layers.css.ts` stylesheet, and use those when creating styles.

```ts compiled
// reset.css.ts

import { style } from '@vanilla-extract/css';
import { reset } from './layers.css.ts';

export const noMargin = style({
  '@layer': {
    [reset]: {
      margin: 0
    }
  }
});

// layers.css.ts

import { globalLayer } from '@vanilla-extract/css';

export const reset = globalLayer('reset');
export const framework = globalLayer('framework');
export const typography = globalLayer(

>   { parent: 'framework' },

Author: I think this should be referencing the variable.

```suggestion
  { parent: framework },
```

  'typography'
);
```

This is particularly useful when using the nested layers feature, as the parent and child names are computed.

In our example, the name of the typography layer becomes `framework.typography`:

```ts compiled
// text.css.ts

import { style } from '@vanilla-extract/css';
import { typography } from './layers.css.ts';

export const standard = style({
  '@layer': {
    [typography]: {
      fontSize: '1rem'
    }
  }
});

// layers.css.ts

import { globalLayer } from '@vanilla-extract/css';

export const reset = globalLayer('reset');
export const framework = globalLayer('framework');
export const typography = globalLayer(

>   { parent: 'framework' },

Author: I think this should be referencing the variable.

```suggestion
  { parent: framework },
```

  'typography'
);
```

## Layer merging

> In order to generate the smallest possible CSS output, Vanilla Extract will merge styles that are assigned to the same layer, if it can be done without impacting the precedence of the rules.

Reviewer: ```suggestion
In order to generate the smallest possible CSS output, Vanilla Extract will merge styles that are assigned to the same layer within the same file, if it can be done without impacting the precedence of the rules.
```

> Notice in this example, while the `themedHeading` style is created before the other styles, it appears later in the stylesheet. This is due to it being assigned to the `theme` layer — which is declared after the `base` layer.
> 
> ```ts compiled
> // typography.css.ts
> import { style, globalLayer } from '@vanilla-extract/css';
> 
> const base = globalLayer('base');
> const theme = globalLayer('theme');
> 
> const themedHeading = style({
>   '@layer': {
>     [theme]: {
>       color: 'rebeccapurple'
>     }
>   }
> });
> const text = style({
>   '@layer': {
>     [base]: {
>       fontSize: '1rem'
>     }
>   }
> });

Author: Tweaked wording slightly to be in line with the latest implementation.

```suggestion
Notice in this example, while the `themedHeading` style is created before the the `heading` style, it appears later in the stylesheet. This is due to it being assigned to the `theme` layer — which is declared after the `base` layer.

```ts compiled
// typography.css.ts
import { style, globalLayer } from '@vanilla-extract/css';

const base = globalLayer('base');
const theme = globalLayer('theme');

const text = style({
  '@layer': {
    [base]: {
      fontSize: '1rem'
    }
  }
});
const themedHeading = style({
  '@layer': {
    [theme]: {
      color: 'rebeccapurple'
    }
  }
});
```

The new output looks like this:

```css
@layer base;
@layer theme;
@layer base {
  .typography_text__94ldqt0 {
    font-size: 1rem;
  }
  .typography_heading__94ldqt2 {
    font-size: 2.4rem;
  }
}
@layer theme {
  .typography_themedHeading__94ldqt1 {
    color: rebeccapurple;
  }
}
```

---
title: globalLayer
parent: global-api
---

# globalLayer

Creates a globally scoped [layer].

```ts compiled
// reset.css.ts

> import { globalLayer, style } from '@vanilla-extract/css';

Reviewer 2: Here, and in a few of the other code blocks, `style` is imported but unused.

Author: I have a TODO item to lint JavaScript code blocks in Markdown using https://github.com/eslint/eslint-plugin-markdown, but I didn't want to overload this already heavy PR.
